02-household-queries/optimize_encoder.py

```diff
@@ -0,0 +1,249 @@
+        index_key="preferredPhrase",
+        chunk_size=chunk_size,
+        chunk_overlap=chunk_overlap,
+        token_limit=token_limit,
+        text_splitter_choice=text_splitter_choice,
+    )
+
+
+def evaluate_retrieval(vectordb):
+    qa = load_training_json()
+    results = []
+    retriever = create_retriever(vectordb)
+    for qa_dict in qa[1:]:
+        orig_question = qa_dict["orig_question"]
+        question = qa_dict.get("question", orig_question)
+        # print(f"\nQUESTION {qa_dict['id']}: {question}")
+        guru_cards = qa_dict.get("guru_cards", [])
+        # print(f"  Desired CARDS : {guru_cards}")
+
+        retrieval = retriever.invoke(question)
+        retrieved_cards = [doc.metadata["source"] for doc in retrieval]
+        results.append(
+            {
+                "id": qa_dict["id"],
+                "question": question,
+                "guru_cards": guru_cards,
+                "retrieved_cards": retrieved_cards,
+                "recall": compute_percent_retrieved(retrieved_cards, guru_cards),
+                "extra_cards": count_extra_cards(retrieved_cards, guru_cards),
+            }
+        )
+    recall_results = print_and_set_recall_stats(results)
+    return recall_results
+
+
+def print_and_set_recall_stats(results):
+    recall_results = {}
+    recall = 0
+    extra_card_count = 0
+
+    print("\nTable:")
+    for res in results:
+        print(res["id"], "|", res["recall"], "|", res["extra_cards"])
+        extra_card_count += res["extra_cards"]
+        recall += res["recall"]
+    recall_results["result"] = results
+    recall_results["recall_percentage"] = recall / len(results)
+    recall_results["extra_card_count"] = extra_card_count
+    return recall_results
+
+
+def run_embedding_func_and_eval_retrieval(
+    embeddings, chunk_size, chunk_overlap, text_splitter_choice
+):
+    selected_embedding = embeddings["func"]
+    persistent_client = chromadb.PersistentClient(
+        settings=Settings(allow_reset=True), path="./chroma_db"
+    )
+
+    vectordb = Chroma(
+        client=persistent_client,
+        collection_name="resources",
+        persist_directory="./chroma_db",
+        embedding_function=selected_embedding,
+    )
+    ingest_call(
+        vectordb=vectordb,
+        chunk_size=chunk_size,
+        chunk_overlap=chunk_overlap,
+        token_limit=embeddings["token_limit"],
+        text_splitter_choice=text_splitter_choice,
+    )
+    recall_results = evaluate_retrieval(vectordb)
+    persistent_client.reset()
+    return recall_results
+
+
+print("""
+      Choose a text splitter
+      1. RecursiveCharacterTextSplitter(Default)
+      2. NLTK
+      3. Spacy
+      """)
+
+text_splitter_choice = input()
+chunk_size = 750
+chunk_overlap = 300
+if text_splitter_choice == "2":
+    nltk.download("punkt")
+elif text_splitter_choice == "3":
+    spacy.cli.download("en_core_web_sm")
+else:
+    print("Define chunk size: 750(default) ")
+    chunk_size = int(input())
```

Comment:
When I press Enter to go with the default, it crashes with:
```
Traceback (most recent call last):
  File "/Users/yoom/dev/labs-gen-ai-experiments/02-household-queries/optimize_encoder.py", line 229, in <module>
    chunk_size = int(input())
                 ^^^^^^^^^^^^
ValueError: invalid literal for int() with base 10: ''
```
Need to provide a `default="750"` argument to `input()`.
Same for the next couple of lines.